marimbabot_audio/scripts/onset_detection.py

```diff
@@ -0,0 +1,547 @@
+        )
+    else:
+        rospy.loginfo("Audio compatible")
+        return True
+    return False
+
+def unpack_data(data):
+    return np.array(struct.unpack(
+        f"{int(len(data) / 2)}h",
+        bytes(data)
+        ),
+        dtype=float)
+
+
+class OnsetDetection:
+    def _init_instrument_config(self):
+        # # harp
+        # self.fmin_note = "C4"
+        # self.fmax_note = "C6"
+        # self.semitones = 62
+
+        # # guzheng
+        # self.fmin_note = "C2"
+        # self.fmax_note = "C8"
+        # self.semitones = 84
+
+        # marimba
+        self.fmin_note = "C4"  # C4 pm_id=60
+        self.fmin_note_id = note_to_pm_id(self.fmin_note)
+        self.fmax_note = "C7"  # C7 pm_id=96
+        self.fmax_note_id = note_to_pm_id(self.fmax_note)
+        self.semitones = 36 + 24  # 60   36 for 4-6 octives, 24 for overtones for two octives
+
+        # convert the western notation to the corresponding frequency
+        self.fmin = note_to_hz(self.fmin_note)
+        self.fmax = note_to_hz(self.fmax_note)
+
+    def _init_audio_config(self):
+        """
+            The configuration of the audio signal
+        """
+        self.sr = 44100
+        self.hop_length = 512
+
+    def _init_detection_config(self):
+        """
+            confidence threshold for note classification(crepe)
+        """
+        self.confidence_threshold = 0.3  # the threshold for note classification
+        self.windows_for_classification = 0.1  # using 0.1 sec data after onset time for note classification
+        # preload model to not block the callback on first message
+        # capacities: 'tiny', 'small', 'medium', 'large', 'full'
+        self.crepe_model = "full"  # choose the crepe model type for music note classification
+        rospy.loginfo(f"Loading crepe {self.crepe_model}-model...")
+        crepe.core.build_and_load_model(self.crepe_model)
+        rospy.loginfo(f"Crepe {self.crepe_model}-model loaded.")
+
+    def _init_visualization_config(self):
+        """
+            config the spectrum plot of
+        """
+        hsv = plt.get_cmap("hsv")
+        self.cmap = ListedColormap(np.vstack((
+            hsv(np.linspace(0, 1, 86)),
+            hsv(np.linspace(0, 1, 85)),
+            hsv(np.linspace(0, 1, 85)))
+        ))
+        self.cmap.set_bad((0, 0, 0, 1))  # make sure they are visible
+
+        # number of samples for analysis window
+        self.window_t = 1.0
+        # and overlap regions between consecutive windows
+        self.window_overlap_t = 0.5
+        self.window_all_t = self.window_t+2*self.window_overlap_t
+
+        self.window = int(self.sr * self.window_t)
+        self.window_overlap = int(self.sr * self.window_overlap_t)
+
+
+        self.overlap_hops = int(self.window_overlap / self.hop_length)
+
+        self.cv_bridge = cv_bridge.CvBridge()
+
+    def _warm_up(self):
+        # warm up classifier / jit caches
+        _ = self.cqt()
```

Comment:
The `_` is not needed afaik. You can just run `self.cqt()`.

Reply:
done